R/read_csv.R

```diff
@@ -0,0 +1,239 @@
+      csv_file_info$inverse_mass_matrix_diag <- rapply(strsplit(gsub("# ", "", line), ","), as.numeric)
+      diagonal_matrix_next <- FALSE
+    }
+  }
+  close(con)
+  if(is.null(csv_file_info$inverse_mass_matrix_diag)) {
+    csv_file_info$inverse_mass_matrix_diag <- NULL
+  }
+  if(csv_file_info$method != "sample") {
+    stop("Supplied CSV file was not generated with sampling. Consider using read_optim_csv or read_vb_csv!")
+  }
+  csv_file_info
+}
+
+#' Reads sampling results from the supplied CSV files. Returns a list
+#' containing sampling arguments, the diagonal of the inverse mass
+#' matrix, the post-warmup samples, the sampling parameters and
+#' optionally samples produced during warmup.
+#'
+#' @param output_files Paths to the CSV files to read.
+#'
+#' @return The list of sampling arguments, the diagonal of the inverse mass
+#' matrix, the post-warmup samples, the sampling parameters and warmup samples
+#' if the run was run with save_warmup = 1.
+#'
+read_sample_csv <- function(output_files) {
+  sampling_info <- NULL
+  inverse_mass_matrix_diag <- c()
+  sampling_params_draws <- c()
+  post_warmup_draws_array <- c()
+  warmup_draws_array <- c()
+  for(output_file in output_files) {
+    # read meta data
+    if (is.null(sampling_info)) {
+      sampling_info <- read_sample_info_csv(output_file)
+      inverse_mass_matrix_diag <- sampling_info$inverse_mass_matrix_diag
+    } else {
+      csv_file_info <- read_sample_info_csv(output_file)
+      # check if sampling info matches
+      check_sampling_csv_info_matches(sampling_info,
+                                  csv_file_info)
+      sampling_info$id <- c(sampling_info$id,
+                            csv_file_info$id)
+      sampling_info$init <- c(sampling_info$init,
+                              csv_file_info$init)
+      sampling_info$random_seed <- c(sampling_info$random_seed,
+                                     csv_file_info$random_seed)
+      sampling_info$output_file <- c(sampling_info$output_file,
+                                     csv_file_info$output_file)
+      sampling_info$output_diagnostic_file <- c(sampling_info$output_diagnostic_file,
+                                                csv_file_info$output_diagnostic_file)
+      sampling_info$output_refresh <- c(sampling_info$output_refresh,
+                                        csv_file_info$output_refresh)
+      inverse_mass_matrix_diag <- cbind(inverse_mass_matrix_diag,
+                                        csv_file_info$inverse_mass_matrix_diag)
+    }
+    # read sampling data
+    if(sampling_info$sample_save_warmup == 1) {
+      num_of_draws <- sampling_info$sample_num_samples + sampling_info$sample_num_warmup
+    } else {
+      num_of_draws <- sampling_info$sample_num_samples
+    }
+    draws <- utils::read.csv(output_file, header = TRUE, comment.char = "#")
+    sampling_params_draws <- rbind(sampling_params_draws, draws[, sampling_info$sampler_params])
```

Comment:
The sampling params are returned in one piece, regardless if we save warmup or not. Is that fine?

Reply:
Not a big deal, but maybe we should split them into warmup and post-warmup for convenience and consistency?

Reply:
Makes sense. Done.

the read_sample_csv now returns 

```R
  list(
    sampling_info = sampling_info,
    inverse_mass_matrix = inverse_mass_matrix,
    warmup = warmup_draws_array,
    post_warmup = post_warmup_draws_array,
    warmup_sampler = warmup_sampling_params_draws,
    post_warmup_sampler = post_warmup_sampling_params_draws
  )
```